### PR TITLE
ArduCopter: add WoW to Copter landing check

### PR DESCRIPTION
**Pull request by TunaLobster**
Should we have a WoW switch, that would be a strong indicator that we are on the ground or not. If the WoW is ambiguous, we have no idea.

### CODE REVIEW

**Comment by TunaLobster** [comment]
I believe this is a step towards fixing #523. @rmackay9

File: ArduCopter/config.h [hunk 1]
```diff
@@ -762,3 +762,7 @@
 #ifndef HAL_FRAME_TYPE_DEFAULT
 #define HAL_FRAME_TYPE_DEFAULT AP_Motors::MOTOR_FRAME_TYPE_X
 #endif
+
+#ifndef LANDING_GEAR_ENABLED
+ #define LANDING_GEAR_ENABLED !HAL_MINIMIZE_FEATURES
```
Comment (khancyr) [inline ArduCopter/config.h:767]: this is wrong as we always have landing gear lib in copter.

File: ArduCopter/land_detector.cpp [hunk 2]
```diff
@@ -76,7 +76,15 @@ void Copter::update_land_detector()
         // if we have a healthy rangefinder only allow landing detection below 2 meters
         bool rangefinder_check = (!rangefinder_alt_ok() || rangefinder_state.alt_cm_filt.get() < LAND_RANGEFINDER_MIN_ALT_CM);
 
-        if (motor_at_lower_limit && accel_stationary && descent_rate_low && rangefinder_check) {
+#if LANDING_GEAR_ENABLED == ENABLED
+        // if we have weight on wheels (WoW) or ambiguis unknown. never no WoW
+        bool WoW_check = (landinggear.get_wow_state() == 1 || landinggear.get_wow_state() == -1) ? true : false;
+#else
+        // no landing gear so cannot hold up the check
```
Comment (khancyr) [inline ArduCopter/land_detector.cpp:83]: comment is wrong
no wow sensor so cannot hold up the check

File: ArduCopter/land_detector.cpp [hunk 3]
```diff
@@ -76,7 +76,15 @@ void Copter::update_land_detector()
         // if we have a healthy rangefinder only allow landing detection below 2 meters
         bool rangefinder_check = (!rangefinder_alt_ok() || rangefinder_state.alt_cm_filt.get() < LAND_RANGEFINDER_MIN_ALT_CM);
 
-        if (motor_at_lower_limit && accel_stationary && descent_rate_low && rangefinder_check) {
+#if LANDING_GEAR_ENABLED == ENABLED
+        // if we have weight on wheels (WoW) or ambiguis unknown. never no WoW
+        bool WoW_check = (landinggear.get_wow_state() == 1 || landinggear.get_wow_state() == -1) ? true : false;
```
Comment (khancyr) [inline ArduCopter/land_detector.cpp:81]: this is cheap call, so I don't think we need the #if . just adding the check should be good.

**Review by khancyr — CHANGES_REQUESTED** [review]
In the current form, this won't improve the landing checking much.
The wow sensor will only allow to confirm that we aren't on the ground preventing earlier landing state.
This won't speed up the landing detection.
Anyway that seems pretty safe to get after removing the #if cause.

File: ArduCopter/config.h (hunk 1 above)
Comment (TunaLobster) [reply ArduCopter/config.h:767]: I was modeling off Plane for this. Dropped this change.

File: ArduCopter/land_detector.cpp (hunk 3 above)
Comment (TunaLobster) [reply ArduCopter/land_detector.cpp:81]: #if and the else to go along with it have been removed.

File: ArduCopter/land_detector.cpp (hunk 2 above)
Comment (TunaLobster) [reply ArduCopter/land_detector.cpp:83]: Removed

**Comment by TunaLobster** [comment]
I didn't want to override with this as something that says guaranteed landing. It looks like the assumption is that the copter is flying unless we are very sure that we are on the ground. While a WoW switch can be a very good indicator of being on the ground, there could still be a false positive from a faulty switch.

**Comment by khancyr** [comment]
@TunaLobster yep no problem ! I made a small analysis for the other. As it is now, it is safe to merge as it shouldn't bother the current behavior.
If we want the switch to be us on priority we would surely need to write up a small algorithm and some wait to test to ensure we limit false positive

**Comment by rmackay9** [comment]
As @khancyr says, the problem is that it won't improve the landing detection because it only adds another condition instead of removing or reducing the importance of others.

I think we would end up with upset users who have gone to the trouble of adding weight-on-wheel detectors and then find it doesn't help so I'm afraid we can't merge this as is.

Land detection is tricky and critical so maybe we can discuss some ideas that would make it work better.  I think the low motor requirement is the one that causes the most trouble although high vibrations can also stop the landing detector from working.  Maybe we could relax the ranges for some of the other conditions.

**Comment by TunaLobster** [comment]
Loosening the other checks if WoW is enabled makes sense. I'm going to say that I don't have the experience to know how much to loosen them. Would you want to adjust the WoW disabled checks in this PR as well?

File: ArduCopter/land_detector.cpp [hunk 4]
```diff
@@ -76,7 +76,10 @@ void Copter::update_land_detector()
         // if we have a healthy rangefinder only allow landing detection below 2 meters
         bool rangefinder_check = (!rangefinder_alt_ok() || rangefinder_state.alt_cm_filt.get() < LAND_RANGEFINDER_MIN_ALT_CM);
 
-        if (motor_at_lower_limit && accel_stationary && descent_rate_low && rangefinder_check) {
+        // if we have weight on wheels (WoW) or ambiguous unknown. never no WoW
+        bool WoW_check = (landinggear.get_wow_state() == 1 || landinggear.get_wow_state() == -1) ? true : false;
```
Comment (peterbarker) [inline ArduCopter/land_detector.cpp:80]:
```suggestion
        bool WoW_check = (landinggear.get_wow_state() == 1 || landinggear.get_wow_state() == -1);
```

**Comment by peterbarker** [comment]
I think it'd be really nice to have WoW as being absolutely definitive "you are on the ground".

Switch failure could be handled with a modification to the WoW functions to use multiple inputs.

The WoW functions should also be modified to use `AP_Button` - that would allow for using more than just logic levels for input in case a switch is marginal under high vibrations.

File: ArduCopter/land_detector.cpp (hunk 4 above)
Comment (TunaLobster) [reply ArduCopter/land_detector.cpp:80]: Could you please explain a little bit what `const` means in this context? To me the value changes and the constant seems odd.

Comment (peterbarker) [reply ArduCopter/land_detector.cpp:80]: `const` simply means the variable can't be modified past its initialisation.  That can help both the compiler and the reader reason about the code.

The ternary operator was just noise - you've got a boolean expression - the ternary operator construct here is a tautology.

**Comment by TunaLobster** [comment]
I would still err that WoW is a strong indicator of being on the ground, but relying on only WoW when we have other methods to check seems...irresponsible.

Is there debounce on anything in AP_LandingGear right now? I see `get_gear_state_duration_ms` and `get_wow_state_duration_ms`, but they are never called anywhere. I also don't see a logic check on `last_wow_event_ms` except in the uncalled function. The original PR for WoW (#7893) had some discussion of 500 ms debounce.

**Comment by peterbarker** [comment]
Closes #523

**Comment by TunaLobster** [comment]
Added a scalar to make the criteria for the accelleration and descent rate less strict with the WoW state is known. I just guessed at making the ranges twice as big. I didn't put the scalar on the range finder check because 2 m is already quite a long ways to the ground. I wasn't sure how to change the criteria for the throttle.

Reworking AP_LandingGear to use AP_Button will probably end up in a separate PR. One problem at a time while I keep learning the code.

File: ArduCopter/land_detector.cpp [hunk 5]
```diff
@@ -67,16 +67,25 @@ void Copter::update_land_detector()
...
         // check that vertical speed is within 1m/s of zero
-        bool descent_rate_low = fabsf(inertial_nav.get_velocity_z()) < 100;
+        bool descent_rate_low = fabsf(inertial_nav.get_velocity_z()) < 100 * land_detector_scalar;
 
         // if we have a healthy rangefinder only allow landing detection below 2 meters
         bool rangefinder_check = (!rangefinder_alt_ok() || rangefinder_state.alt_cm_filt.get() < LAND_RANGEFINDER_MIN_ALT_CM);
 
-        if (motor_at_lower_limit && accel_stationary && descent_rate_low && rangefinder_check) {
+        // if we have weight on wheels (WoW) or ambiguous unknown. never no WoW
+        const bool WoW_check = (landinggear.get_wow_state() == 1 || landinggear.get_wow_state() == -1);
```
Comment (tridge) [inline ArduCopter/land_detector.cpp:86]: should use enum not 1 and -1

**Review by tridge — CHANGES_REQUESTED** [review]

File: ArduCopter/land_detector.cpp [hunk 6]
```diff
@@ -67,16 +67,25 @@ void Copter::update_land_detector()
         bool motor_at_lower_limit = motors->limit.throttle_lower && attitude_control->is_throttle_mix_min();
 #endif
 
+        uint8_t land_detector_scalar = 1;
+        if (landinggear.get_wow_state() != -1) {
```
Comment (tridge) [inline ArduCopter/land_detector.cpp:71]: use enum

File: ArduCopter/land_detector.cpp (hunk 5 above)
Comment (TunaLobster) [reply ArduCopter/land_detector.cpp:86]: Changed. Still learning syntax.

File: ArduCopter/land_detector.cpp (hunk 6 above)
Comment (TunaLobster) [reply ArduCopter/land_detector.cpp:71]: Changed.

**Comment by rmackay9** [comment]
Merged, thanks!

**Pull request merged**